### PR TITLE
Fix issue 3426, Fix validating if swapchain image was acquired

### LINKED ISSUE #3426: NonAcquiredSwapchainImageUsed when recording cmdbuffer prior to acquiring image and submitting cmdbuffer
**Describe the Issue**
With the current tip 4defcad9cd878785d4df4b708194b6a10cb09559 of the Vulkan-ValidationLayers repo built locally I am getting the following error when recording a command buffer that only ever needs to be recorded when swapchain images change (i.e. when creating or resizing the window).

That validation is issued by `CoreChecks::ValidateImageBarrier` -> `ValidateImageAcquired` when recording a `vkCmdPipelineBarrier` with a `VkImageMemoryBarrier` referencing a swapchain `VkImage`. However, in the application I am submitting that command buffer of course only _after_ having acquired the respective swapchain image. But the validation thinks that you are going to record a command buffer per frame _after_ you have acquired the swapchain image. Is that right?

**Valid Usage ID**
```
Validation Error: [ UNASSIGNED-CoreValidation-NonAcquiredSwapchainImageUsed ] Object 0: handle = 0xf56c9b0000000004, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0xe8282151 | vkCmdPipelineBarrier: Image VkImage 0xf56c9b0000000004[] is a presentable image, but it is currently not acquired from swapchain.
```

**Environment:**
 - OS: Windows x64 21H1
 - GPU: Nvidia GeForce RTX 2080 Ti
 - SDK or header version if building from repo: 4defcad9cd878785d4df4b708194b6a10cb09559
 - Options enabled (synchronization, best practices, etc.): none / defaults

**Additional context**
none

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes #3426

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -8485,6 +8485,21 @@ bool CoreChecks::PreCallValidateCmdPipelineBarrier2KHR(VkCommandBuffer commandBu
...
+void CoreChecks::EnqueueVerifyPipelineBarrier(VkCommandBuffer command_buffer, uint32_t imageMemoryBarrierCount,
+                                              const VkImageMemoryBarrier *pImageMemoryBarriers, const char *func_name) {
+    CMD_BUFFER_STATE *cb_state = GetCBState(command_buffer);
+
+    // Enqueue the submit time validation here, ahead of the submit time state update in the StateTracker's PostCallRecord
+    cb_state->barrierUpdates.emplace_back([this, imageMemoryBarrierCount, pImageMemoryBarriers, func_name]() {
+        bool skip = false;
+        for (uint32_t i = 0; i < imageMemoryBarrierCount; ++i) {
+            const auto image_state = GetImageState(pImageMemoryBarriers[i].image);
+            skip |= ValidateImageAcquired(*image_state, func_name);
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:8497]: Should we check if `image_state` is not null here? e.g., could a user pass in an image that was previously destroyed to a barrier call without getting an error previously? I don't think they can, but not 100% sure (just from a quick look, looks like validation is skipped in `ValidateImageBarrier` if `image_state` is null).

**Review by ncesario-lunarg — COMMENTED** [review]
One question, but I think this looks good.

FYI @jeremyg-lunarg for the addition to `CMD_BUFFER_STATE`.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -8485,6 +8485,20 @@ bool CoreChecks::PreCallValidateCmdPipelineBarrier2KHR(VkCommandBuffer commandBu
     return skip;
 }
 
+void CoreChecks::EnqueueVerifyPipelineBarrier(VkCommandBuffer command_buffer, uint32_t imageMemoryBarrierCount,
+                                              const VkImageMemoryBarrier *pImageMemoryBarriers, const char *func_name) {
+    CMD_BUFFER_STATE *cb_state = GetCBState(command_buffer);
+
+    // Enqueue the submit time validation here, ahead of the submit time state update in the StateTracker's PostCallRecord
+    cb_state->barrierUpdates.emplace_back([this, imageMemoryBarrierCount, pImageMemoryBarriers, func_name]() {
```
Comment (Tony-LunarG) [inline layers/core_validation.cpp:8493]: Is pImageMemoryBarriers safe to emplace here? I think it's pointing to app data.

**Review by Tony-LunarG — CHANGES_REQUESTED** [review]

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -8485,6 +8485,24 @@ bool CoreChecks::PreCallValidateCmdPipelineBarrier2KHR(VkCommandBuffer commandBu
     return skip;
 }
 
+void CoreChecks::EnqueueVerifyPipelineBarrier(VkCommandBuffer command_buffer, uint32_t imageMemoryBarrierCount,
+                                              const VkImageMemoryBarrier *pImageMemoryBarriers, const char *func_name) {
+    CMD_BUFFER_STATE *cb_state = GetCBState(command_buffer);
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:8490]: Should this function return if `imageMemoryBarrierCount == 0`?

File: layers/queue_state.h [hunk 4]
```diff
@@ -71,6 +73,15 @@ class SEMAPHORE_STATE : public REFCOUNTED_NODE {
           payload(type_create_info ? type_create_info->initialValue : 0) {}
 
     VkSemaphore semaphore() const { return handle_.Cast<VkSemaphore>(); }
+
+    void execute_waiting_functions() {
```
Comment (jeremyg-lunarg) [inline layers/queue_state.h:77]: shouldn't this be ExecuteWaitingFunctions()?

**Comment by ziga-lunarg** [comment]
> I'm getting a crash running the surge in LvlFindInChain here:
> 
> ```
> void ValidationStateTracker::PostCallRecordQueuePresentKHR(VkQueue queue, const VkPresentInfoKHR *pPresentInfo, VkResult result) {
>     // Semaphore waits occur before error generation, if the call reached the ICD. (Confirm?)
>     const auto present = [this, pPresentInfo, result]() {
>         const auto *present_id_info = LvlFindInChain<VkPresentIdKHR>(pPresentInfo->pNext);
> ```

Do you have a call stack?

File: layers/queue_state.h [hunk 5]
```diff
@@ -71,6 +73,15 @@ class SEMAPHORE_STATE : public REFCOUNTED_NODE {
           payload(type_create_info ? type_create_info->initialValue : 0) {}
 
     VkSemaphore semaphore() const { return handle_.Cast<VkSemaphore>(); }
+
+    void ExecuteWaitingFunctions() {
+        for (auto& func : waiting_functions) {
+            if (func.use_count() == 1) {
```
[the remark this answers is not in the document]
Comment (ziga-lunarg) [reply layers/queue_state.h:79]: `vkQueuePresentKHR()` can wait on multiple semaphores, only the last semaphore that gets signaled should execute this function.

[the remark this answers is not in the document]
Comment (ncesario-lunarg) [reply layers/queue_state.h:79]: I guess I'm having trouble seeing how this condition guarantees only the last one gets called. If `func` was added `N` times to  `waiting_functions`, wouldn't `use_count` be `N` for the duration of this loop? Is `func.reset()` meant to be called after `(*func)();`?

[the remark this answers is not in the document]
Comment (ncesario-lunarg) [reply layers/queue_state.h:79]: :+1: thanks, think I finally got it :)

File: layers/image_state.h [hunk 6]
```diff
@@ -229,10 +229,16 @@ class IMAGE_VIEW_STATE : public BASE_NODE {
     VkExtent3D GetExtent() const;
 };
 
+enum SwapchainImageState {
+    NON_ACQUIRED,
+    WAITING_SEMAPHORE,
```
Comment (ncesario-lunarg) [inline layers/image_state.h:234]: Is this used anywhere?

**Pull request closed without merging**